src/main/kotlin/com/openlattice/data/ids/jobs/FixDuplicateIdAssignmentJob.kt

```diff
@@ -0,0 +1,301 @@
+        }) { rs ->
+            ResultSetAdapters.id(rs) to mapper.readValue<List<EntityKey>>(rs.getString(DUPLICATES_FIELD))
+        }.toMap()
+
+        entityKeysAndIds
+                .asSequence()
+                .forEach { (id, entityKeys) ->
+                    val entitySetAppearances = entityKeys.groupingBy { it.entitySetId }.eachCount()
+
+                    if (entitySetAppearances.values.all { it == 1 }) {
+                        /**
+                         * Each entity set appears only once. We will assign new ids to the entity keys and
+                         */
+                        assignNewIdsAndMoveData(id, entityKeys.subList(1, entityKeys.size))
+
+                    } else {
+                        /**
+                         * We have two entity keys in the same entity set assigned to the same id. This is
+                         * extra special bad and we do not know how to handle without manual intervention
+                         */
+                        logger.error("Woe is us, we will have to manually address.")
+                        recordCollision(id, entityKeys)
+                    }
+                }
+
+        state.idsProceessed += entityKeysAndIds.values.sumBy { it.size }
+        state.id = ids.max() ?: throw IllegalStateException("Entity key ids cannot be empty.")
+        hasWorkRemaining = (ids.size == BATCH_SIZE)
+    }
+
+    private fun assignNewIdsAndMoveData(originalId: UUID, entityKeys: List<EntityKey>) {
+        val newIds = idService.getNextIds(entityKeys.size).iterator()
+        hds.connection.use { connection ->
+            connection.autoCommit = false
+
+            val psSyncs = connection.prepareStatement(UPDATE_ID_ASSIGNMENT)
+            val psIds = connection.prepareStatement(MIGRATE_IDS)
+            val psData = connection.prepareStatement(MIGRATE_DATA)
+            val psEdges = connection.prepareStatement(MIGRATE_EDGES)
+            val psDst = connection.prepareStatement(UPDATE_E_DST)
+            val psEdge = connection.prepareStatement(UPDATE_E_EDGE)
+
+            entityKeys.forEach { entityKey ->
+                val newId = newIds.next()
+                updateIdAssignment(psSyncs, entityKey, newId)
+                migrate(psIds, entityKey.entitySetId, originalId)
+                migrate(psData, entityKey.entitySetId, originalId)
+                migrate(psEdges, entityKey.entitySetId, originalId)
+                update(psDst, newId, originalId)
+                update(psEdge, newId, originalId)
+            }
+            listOf(psSyncs,psIds, psData, psEdges, psDst, psEdge).forEach { it.executeBatch() }
+
+            connection.commit()
+        }
+    }
+
+    private fun migrate(
+            ps: PreparedStatement,
+            originalId: UUID,
+            entitySetId: UUID
+    ) {
+        bind(ps, entitySetId, originalId)
+        ps.addBatch()
+    }
+
+    private fun update(
```

Comment:
just double checking, `migrate` and `update` are identical and just separated because the arguments are semantically different, yeah?

Reply:
yes